### PR TITLE
[RW-9656][risk=low][part 1] of firecloud-orch deprecation - move billing API

### PR DESCRIPTION
**Pull request by yonghaoy**
Description:
FixDesynchronizedBillingProjectOwners is still using firecloud-orch one, will do that in pt2 pr, it is not working anyway...

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 1]
```diff
@@ -243,52 +241,46 @@ public String createAllOfUsBillingProject(String billingProjectName, String serv
...
   public void deleteBillingProject(String billingProjectName) {
     BillingV2Api billingV2Api = serviceAccountBillingV2ApiProvider.get();
-    retryHandler.run(
+    rawlsRetryHandler.run(
         (context) -> {
-          billingV2Api.deleteBillingProject(billingProjectName);
+          billingV2Api.deleteBillingProjectV2(billingProjectName);
           return null;
         });
   }
```
Comment (Peter-Lavigne) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:249]: I think methods like this that only use Rawls endpoints should be moved into a Rawls service, since they are no longer relying on firecloud-orchestration.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java [hunk 2]
```diff
@@ -303,22 +295,17 @@ public void removeOwnerFromBillingProject(
...
-    final BillingApi scopedBillingApi;
+    final BillingV2Api billingV2Api;
 
     if (callerAccessToken.isPresent()) {
-      // use a private instance of BillingApi instead of the provider
-      // b/c we don't want to modify its ApiClient globally
-      final ApiClient apiClient = firecloudApiClientFactory.newApiClient();
-      apiClient.setAccessToken(callerAccessToken.get());
-      scopedBillingApi = new BillingApi();
-      scopedBillingApi.setApiClient(apiClient);
```
Comment (Peter-Lavigne) [inline api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:314]: Why was this removed? It seems like the original intent was to allow overriding the access token, but now the access token override value is unused.

**Review by Peter-Lavigne — CHANGES_REQUESTED** [review]
Looks mostly solid with a few suggestions.

However, I'm not sure how to test this. I was trying to log outgoing API requests but I'm not sure how to configure the app to provide those logs. How did you test this?

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
+package org.pmiops.workbench.firecloud;
```
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java:1]: the issues is we have too many packages/directories. The way we organize packages is messy and problematic. 
I was thinking using `firecloud` dir for most firecloud services, instead of kept of creating new ones. Just like we did for `google` package. There would be few more ApiClients coming to that dir soon, those would be small, and I don't think it works to have a different directories.

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  @RequestScope(proxyMode = ScopedProxyMode.DEFAULT)
+  public WorkspacesApi workspacesApi(@Qualifier(END_USER_API_CLIENT) ApiClient apiClient) {
+    WorkspacesApi api = new WorkspacesApi();
+    api.setApiClient(apiClient);
+    return api;
+  }
+
+  @Bean(name = END_USER_LENIENT_TIMEOUT_WORKSPACE_API)
+  @RequestScope(proxyMode = ScopedProxyMode.DEFAULT)
+  public WorkspacesApi lenientTimeoutWorkspacesApi(
```
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java:89]: okay, removing them and just kept billing. 
it would easier to just copy everything over... So I did that.

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+public class RawlsApiClientFactory {
+  public static final String X_APP_ID_HEADER = "X-App-ID";
+
+  private static final String ADMIN_SERVICE_ACCOUNT_NAME = "firecloud-admin";
+  // The set of Google OAuth scopes required for access to FireCloud APIs. If FireCloud ever changes
+  // its API scopes (see https://api.firecloud.org/api-docs.yaml), we'll need to update this list.
+  public static final List<String> SCOPES =
+      ImmutableList.of(
+          "https://www.googleapis.com/auth/userinfo.profile",
+          "https://www.googleapis.com/auth/userinfo.email");
```
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java:28]: I copied from Firecloud codebase, if if works in firecloud, it should work in Rawls in the same way. 
Where did we see require open id or billing scope in our codebase?

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  @RequestScope(proxyMode = ScopedProxyMode.DEFAULT)
+  public ApiClient endUserLenientTimeoutApiClient(
+      UserAuthentication userAuthentication,
+      RawlsApiClientFactory factory,
+      WorkbenchConfig config) {
+    ApiClient apiClient = factory.newRawlsApiClient();
+    apiClient.setAccessToken(userAuthentication.getCredentials());
+    apiClient
+        .getHttpClient()
+        .setReadTimeout(config.firecloud.lenientTimeoutInSeconds, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java:63]: Workspace API 

> 



> Also, thank you for splitting this into another PR! I personally found this one much easier to review, since the complexity of the migrating all the endpoints is now isolated to the next PR.

NP! It is definitely to review, but it does add extra couple hours of splitting, rewriting some, then future solve conflift.
I underestimate the error so did not consider start from small. But actually I still prefer to keep that large PR since it is already there, and there is choice between 5+ extra hours of rewriting vs 1 extra hr reviewing code given the extra logic are really just fixing tests....

But anyway, it is good to hear spliting makes it easier to review.

**Comment by yonghaoy** [comment]
> Looks mostly solid with a few suggestions.
> 
> However, I'm not sure how to test this. I was trying to log outgoing API requests but I'm not sure how to configure the app to provide those logs. How did you test this?


I  start local API server and test CRUD workspace. Make sure they are still working. 
That should cover this PR and the next one.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 2 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:314]: My understanding of this code is if callerAccessToken is present that means the call is from user. In this case we should just use `endUserApi`. Otherwise, we use `serviceAccountApi`.
Probably this code was writing before we invite those two type of API provider.

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+   */
+  public ApiClient newRawlsApiClient() {
+    WorkbenchConfig workbenchConfig = workbenchConfigProvider.get();
+    ApiClient apiClient = new ApiClient();
+    apiClient.setBasePath(workbenchConfig.firecloud.baseUrl);
+    apiClient.addDefaultHeader(X_APP_ID_HEADER, workbenchConfig.firecloud.xAppIdValue);
+    apiClient.setDebugging(workbenchConfig.firecloud.debugEndpoints);
+    apiClient
+        .getHttpClient()
+        .setReadTimeout(workbenchConfig.firecloud.timeoutInSeconds, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java:97]: ah you are right.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 1 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:249]: ditto to other fireclou vs rawls comment. Happy to chat offline about this topic.

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsConfig.java:62]: Fair, and I think this is still an improvement as-is 👍

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java:28]: It's defined in Rawls' OpenAPI spec. I suspect it's working today because we're also passing those additional scopes.

[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java:28]: But we don't explicitly require those scopes here, so it's possible that it would unexpectedly break if we stopped passing those scopes.

[the remark this answers is not in the document]
Comment (Peter-Lavigne) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java:28]: "It works today" is valid for a refactoring PR, though, so I'm OK with this.

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 2 above)
Comment (Peter-Lavigne) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:314]: Interestingly, our only use of `removeOwnerFromBillingProject` passes `Optional.empty()`. I think it'd be best to remove the parameter and logic entirely since we don't need it.

**Comment by Peter-Lavigne** [comment]
> I start local API server and test CRUD workspace. Make sure they are still working.

How did you validate that our API is using Rawls instead of firecloud-orchestration?

File: api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/RawlsApiClientFactory.java:28]: ah, i saw billing scopes was in `FirecloudConfig` `RawlsConfig`. Not sure how those are organized, looks a bit wired. I probably will have couple followup to clean up things, consolidate those classes after this is done. 
 https://github.com/all-of-us/workbench/blob/ad57c9f24861b27d94e3dbd6ab9b9f776d95e34f/api/src/main/java/org/pmiops/workbench/firecloud/FireCloudConfig.java#L37

File: api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java (hunk 2 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/firecloud/FireCloudServiceImpl.java:314]: changed to `removeOwnerFromBillingProjectAsService`

**Comment by yonghaoy** [comment]
> > I start local API server and test CRUD workspace. Make sure they are still working.
> 
> How did you validate that our API is using Rawls instead of firecloud-orchestration?

that is good point. I don't know. Maybe checking firecloud-orch logs?

**Comment by Peter-Lavigne** [comment]
Yonghao was able to test this by forcing an error to occur and checking that the logs indicate that Rawls is being used:

```
[Thu Apr 20 13:15:49 EDT 2023] WARNING: org.pmiops.workbench.utils.ResponseCodeRetryPolicy logRetry - Rawls API unavailable, retrying after 1 attempts 
org.pmiops.workbench.rawls.ApiException: Internal Server Error
        at org.pmiops.workbench.rawls.ApiClient.handleResponse(ApiClient.java:900)
        at org.pmiops.workbench.rawls.ApiClient.execute(ApiClient.java:816)
        at org.pmiops.workbench.rawls.ApiClient.execute(ApiClient.java:799)
        at org.pmiops.workbench.rawls.api.BillingV2Api.deleteBillingProjectV2WithHttpInfo(BillingV2Api.java:773)
        at org.pmiops.workbench.rawls.api.BillingV2Api.deleteBillingProjectV2(BillingV2Api.java:750)
        at org.pmiops.workbench.firecloud.FireCloudServiceImpl.lambda$deleteBillingProject$3(FireCloudServiceImpl.java:246)
```

**Review by Peter-Lavigne — APPROVED** [review]
